nipype/interfaces/base/traits_extension.py

```diff
@@ -465,3 +463,73 @@ class InputMultiObject(MultiObject):
 
 InputMultiPath = InputMultiObject
 OutputMultiPath = OutputMultiObject
+
+
+def _rebase_path(value, cwd):
+    if isinstance(value, list):
+        return [_rebase_path(v, cwd) for v in value]
+
+    try:
+        value = Path(value)
+    except TypeError:
+        pass
+    else:
+        try:
+            value = value.relative_to(cwd)
+        except ValueError:
+            pass
+    return value
+
+
+def _resolve_path(value, cwd):
+    if isinstance(value, list):
+        return [_resolve_path(v, cwd) for v in value]
+
+    try:
+        value = Path(value)
+    except TypeError:
+        pass
+    else:
+        if not value.is_absolute():
+            value = Path(cwd).absolute() / value
+    return value
+
+
+def _recurse_on_path_traits(func, thistrait, value, cwd):
+    """Run func recursively on BasePath-derived traits."""
+    if thistrait.is_trait_type(BasePath):
+        value = func(value, cwd)
+    elif thistrait.is_trait_type(traits.List):
+        innertrait, = thistrait.inner_traits
+        if not isinstance(value, (list, tuple)):
+            value = [value]
+
+        value = [_recurse_on_path_traits(func, innertrait, v, cwd)
+                 for v in value]
+    elif isinstance(value, dict) and thistrait.is_trait_type(traits.Dict):
+        _, innertrait = thistrait.inner_traits
+        value = {k: _recurse_on_path_traits(func, innertrait, v, cwd)
+                 for k, v in value.items()}
+    elif isinstance(value, tuple) and thistrait.is_trait_type(traits.Tuple):
+        value = tuple([_recurse_on_path_traits(func, subtrait, v, cwd)
+                       for subtrait, v in zip(thistrait.handler.types, value)])
+    elif thistrait.is_trait_type(traits.TraitCompound):
+        is_str = [isinstance(f, (traits.String, traits.BaseStr, traits.BaseBytes, Str))
+                  for f in thistrait.handler.handlers]
+        if any(is_str) and isinstance(value, (bytes, str)) and not value.startswith('/'):
+            return value
+
+        for subtrait in thistrait.handler.handlers:
+            value = _recurse_on_path_traits(func, subtrait(), value, cwd)
```

Comment:
Oh, I see. We're updating `value` each time, so as long as one is a Path-like, then it gets rebased/resolved.

Do we show idempotence? e.g.,

```Python
for func in _rebase_path, _resolve_path:
    newval = _recurse_on_path_traits(func, thistrait, value, cwd)
    assert newval == _recurse_on_path_traits(func, thistrait, newval, cwd)
```

I looked for it in the tests, but you may have had this in mind already but written it in a way I didn't recognize.

Reply:
Please let me know if the latest commit addresses this issue

Reply:
Hard to tell... it's not the most readable test, so I'm working at trying to figure out if there's an idempotence test in there somewhere.

Reply:
It doesn't look like it. I see a lot of inverse tests, e.g., f(f^-1(x)) = x and f^-1(f(y)) = y. In case it wasn't clear, by idempotence, I mean f(f(x)) = f(x). In case it was clear, could you clearly label those tests?

Reply:
Example:

https://github.com/nipy/nipype/pull/2970/files#diff-2f70a0b8533ee2d429dcdfd7d1bcec87R39

It uses the output from the rebase function. It is true that I'm not testing idempotence bidirectionally [I'm testing x = f'(f(x)) but I'm not testing y = f(f'(y))]

Reply:
Okay, I believe now I made idempotence tests :)